### PR TITLE
4750-Message-list-migration

### PR DESCRIPTION
**Pull request by latin-panda**
# Description

This PR will add the migration from messages tab's list with the required services and redux logic.
It will also add the configuration for unit test in karma using mocha, chai and sinon.

medic/cht-core#4750

# Code review items

- Readable: Concise, well named, follows the [style guide](https://docs.communityhealthtoolkit.org/contribute/code/style-guide/), documented if necessary.
- Documented: Configuration and user documentation on [cht-docs](https://github.com/medic/cht-docs/)
- Tested: Unit and/or e2e where appropriate
- Internationalised: All user facing text
- Backwards compatible: Works with existing data and configuration or includes a migration. Any breaking changes documented in the release notes.

# License

The software is provided under AGPL-3.0. Contributions to this project are accepted under the same license.

### CODE REVIEW

File: webapp/src/ts/reducers/messages.ts [hunk 1]
```diff
@@ -0,0 +1,30 @@
+import { createReducer, on } from '@ngrx/store';
+
+import { Actions } from '../actions/messages';
+
+export interface MessagesState {
+  error: boolean;
+  conversations: object[],
```
Comment (latin-panda) [inline webapp/src/ts/reducers/messages.ts:7]: I'll add proper typing in next PR when loading message content and have better idea of the fields.

File: webapp/package.json [hunk 2]
```diff
@@ -63,5 +63,19 @@
...
+    "chai": "^4.2.0",
+    "karma": "~5.0.0",
+    "karma-chai": "^0.1.0",
+    "karma-chrome-launcher": "~3.1.0",
+    "karma-mocha": "^2.0.1",
+    "karma-mocha-reporter": "^2.2.5",
+    "karma-sinon": "^1.0.5",
+    "mocha": "^8.1.3",
+    "sinon": "^9.0.3"
+  }
```
Comment (garethbowen) [inline webapp/package.json:80]: We already have most of these dependencies at the root level which is why the `devDependencies` was empty before. I think you'll be able to find a way to use the root dependencies rather than having to have them in the `webapp/package.json` too.

File: webapp/src/ts/modules/messages/messages.component.html [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+            <p *ngIf="error" class="alert alert-danger" role="alert" translate>Error fetching messages</p>
+            <p class="loading-status" *ngIf="!error && !loading && !conversations.length" translate>No messages found</p>
+            <p class="loading-status" *ngIf="!error && !loading && conversations.length">{{ 'message.list.complete' | translate }}</p>
+            <div class="loader" *ngIf="loading"></div>
+            <div class="padding"></div>
+        </div>
+
+        <router-outlet></router-outlet>
+    </div>
+</div>
```
Comment (garethbowen) [inline webapp/src/ts/modules/messages/messages.component.html:40]: Indentation is all off - make sure your editor is configured correctly.

File: webapp/src/ts/modules/messages/messages.routes.ts [hunk 4]
```diff
@@ -0,0 +1,13 @@
...
+import { MessagesComponent } from './messages.component';
+
+export const routes: Routes = [
+    {
+        path: 'messages',
+        component: MessagesComponent,
+        data: { permissions: [ 'can_view_messages' ] },
+        canActivate: [ RouteGuardProvider ],
+    },
+];
```
Comment (garethbowen) [inline webapp/src/ts/modules/messages/messages.routes.ts:13]: Indentation again.

File: webapp/src/ts/reducers/messages.ts (hunk 1 above)
Comment (garethbowen) [reply webapp/src/ts/reducers/messages.ts:7]: Remember to only do the bare minimum to replace AngularJS - we can fix typing and other improvements in a future release.

File: webapp/src/ts/services/hydrate-messages.service.ts [hunk 5]
```diff
@@ -0,0 +1,99 @@
+import { Injectable } from '@angular/core';
+import * as _ from 'lodash-es';
```
Comment (garethbowen) [inline webapp/src/ts/services/hydrate-messages.service.ts:2]: Can we be specific and only import functions we use? That way tree-shaking will drop unused functions and save some bundle size automatically.

File: Gruntfile.js [hunk 6]
```diff
@@ -601,6 +601,24 @@ module.exports = function(grunt) {
...
+      },
+      'unit-webapp-continuous': {
+        cmd: () => {
+          return [
+            `cd webapp`,
+            '../node_modules/.bin/ng test webapp --watch=true',
+            'cd ../',
+          ].join(' && ');
+        }
+      },
```
Comment (garethbowen) [inline Gruntfile.js:621]: I feel like there should be a better way, but as long as this runs it'll do for now and we can figure out the "right" way later.

**Review by garethbowen — CHANGES_REQUESTED** [review]
Great effort! Nice one.

I've left a few suggestions inline. I don't know Angular very well so I'll also tag Diana to review.

It looks like CI is failing on eslint, but we can polish that up later.

File: webapp/package.json (hunk 2 above)
Comment (latin-panda) [reply webapp/package.json:80]: Ohhh I see, let me try using the root dependencies and I'll get back on that.

File: webapp/src/ts/modules/messages/messages.routes.ts (hunk 4 above)
Comment (latin-panda) [reply webapp/src/ts/modules/messages/messages.routes.ts:13]: Thanks! My IDE isn't picking up the eslint for some reason. Fixed

File: webapp/src/ts/modules/messages/messages.component.html (hunk 3 above)
Comment (latin-panda) [reply webapp/src/ts/modules/messages/messages.component.html:40]: Fixed

File: Gruntfile.js (hunk 6 above)
Comment (latin-panda) [reply Gruntfile.js:621]: Yes, agree, maybe we can create a task for further research. We need the angular cli for running unit test.

File: webapp/src/ts/services/hydrate-messages.service.ts (hunk 5 above)
Comment (latin-panda) [reply webapp/src/ts/services/hydrate-messages.service.ts:2]: Sure, we can try: `import { find as _find, map as _map } from 'lodash-es';`
prefixing with underscore `_find` so that we know this is coming from lodash, specially because `find` and `map` are very generic names.

File: webapp/package.json (hunk 2 above)
Comment (latin-panda) [reply webapp/package.json:80]: I removed dependencies in this file and it's working nice :)

File: webapp/tests/karma/ts/services/hydrate-messages.spec.ts [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+      key: contact._id,
+      contact: contact.name,
+      lineage: _.map(lineage, 'name'),
+      outgoing: true,
+      from: contact._id,
+      date: given[0].value.date,
+      type: 'contact',
+      message: doc.tasks[0].messages[0].message
+    }];
+    lineageModelGeneratorService.reportSubjects.returns(Promise.resolve([{
```
Comment (dianabarsan) [inline webapp/tests/karma/ts/services/hydrate-messages.spec.ts:69]: `lineageModelGeneratorService.reportSubjects.resolves(<value>)`

File: webapp/tests/karma/ts/services/message-contact.spec.ts [hunk 8]
```diff
@@ -0,0 +1,292 @@
...
+      dbService.get = () => ({
+        query,
+        allDocs: sinon.stub().returns(Promise.resolve({}))
+      });
+      hydrateMessagesService.hydrate.returns(Promise.resolve([]));
+
+      return service
+        .getConversation('abc', 45, 120)
+        .then(result => {
+        expect(query.args[0][1]).to.deep.equal(expectedQueryParams);
```
Comment (dianabarsan) [inline webapp/tests/karma/ts/services/message-contact.spec.ts:207]: indentation is off here

File: webapp/src/ts/modules/messages/messages.component.ts [hunk 9]
```diff
@@ -0,0 +1,125 @@
...
+  ngOnInit(): void {
+    this.updateConversations();
+    this.watchForChanges();
+  }
+
+  ngOnDestroy(): void {
+    this.subscriptions.unsubscribe();
+  }
+
+  watchForChanges() {
```
Comment (dianabarsan) [inline webapp/src/ts/modules/messages/messages.component.ts:62]: Shouldn't this be private?

**Review by dianabarsan — CHANGES_REQUESTED** [review]
Great work! I left a couple of comments inline, but nothing major. 

I have a more broad comment about method tagging. I think we should tag all methods (in services, components etc) that we don't mean to "expose" as private. I think it helps in making what the thing should expose more obvious.

File: webapp/tests/karma/ts/services/message-contact.spec.ts (hunk 8 above)
Comment (latin-panda) [reply webapp/tests/karma/ts/services/message-contact.spec.ts:207]: Ah this indentation thing gave me troubles this time. Fixed

File: webapp/tests/karma/ts/modules/messages/messages.component.spec.ts [hunk 10]
```diff
@@ -0,0 +1,174 @@
...
+describe('Messages Component', () => {
+  let component: MessagesComponent;
+  let fixture: ComponentFixture<MessagesComponent>;
+  let store: MockStore;
+  let messageContactService;
+  let changesService;
+
+  beforeEach(async(() => {
+    const messageContactServiceMock = {
+      getList: sinon.stub().returns(Promise.resolve([]))
```
[the remark this answers is not in the document]
Comment (latin-panda) [reply webapp/tests/karma/ts/modules/messages/messages.component.spec.ts:20]: Good one. Fixed

File: webapp/tests/karma/ts/services/hydrate-messages.spec.ts (hunk 7 above)
Comment (latin-panda) [reply webapp/tests/karma/ts/services/hydrate-messages.spec.ts:69]: Fixed

File: webapp/tests/karma/ts/modules/messages/messages.component.spec.ts [hunk 11]
```diff
@@ -0,0 +1,174 @@
...
+      subscribe: sinon.stub().returns(Promise.resolve(of({})))
+    };
+    const mockedSelectors = [
+      { selector: 'getSelectedConversation', value: {} },
+      { selector: 'getConversations', value: []},
+      { selector: 'getLoadingContent', value: false },
+      { selector: 'getMessagesError', value: false },
+    ];
+
+    TestBed.configureTestingModule({
```
[the remark this answers is not in the document]
Comment (latin-panda) [reply webapp/tests/karma/ts/modules/messages/messages.component.spec.ts:32]: Yes, like creating a module with dependencies for testing.

File: webapp/src/ts/actions/messages.ts [hunk 12]
```diff
@@ -0,0 +1,21 @@
+import { Store } from '@ngrx/store';
+
+import { createSingleValueAction } from './actionUtils';
+
+export const Actions = {
+  addSelectedConversation: createSingleValueAction('SET_SELECTED_CONVERSATION', 'selected'),
```
[the remark this answers is not in the document]
Comment (latin-panda) [reply webapp/src/ts/actions/messages.ts:6]: I think I started by naming it Set and then switch it to Add. I have refactored it to Set.

File: webapp/src/ts/reducers/messages.ts [hunk 13]
```diff
@@ -0,0 +1,30 @@
...
+}
+
+const initialState: MessagesState = {
+  error: false,
+  conversations: [],
+  messages: [],
+  selected: null,
+};
+
+const insertConversations = (state, newConversions) => {
```
[the remark this answers is not in the document]
Comment (latin-panda) [reply webapp/src/ts/reducers/messages.ts:19]: yes, and also it'll match to the actions and will be easier to understand. Fixed.

File: webapp/src/ts/modules/messages/messages.component.ts [hunk 14]
```diff
@@ -0,0 +1,125 @@
...
+
+    return this.messageContactService
+      .getList()
+      .then(conversations => {
+        this.setConversations(conversations, { merge });
+        this.loading = false;
+      });
+  }
+
+  removeDeleted(currentConversations = [], updatedConversations = []) {
```
[the remark this answers is not in the document]
Comment (latin-panda) [reply webapp/src/ts/modules/messages/messages.component.ts:99]: Yes. This is simple case, but personally, I've never been fan of access modifiers 🙁 unless it's for critical code. I think they make the testing more complex as we try to reach to those private functions, it's more difficult to see if the private functions got actually tested (we don't have coverage report yet), and if something fails then it requires more effort to detect where the problem is (I experienced it in services). 
Instead, I prefer public or package access (not sure, if js has package) and focus on coverage testing of single unit of code in more controlled and targeted way, that might also lead to more frequent usage of pure functions, so devs are more conscious (responsibility) of the inputs and outputs of the func. Anyhow, I wanted to externalise my unconventional point of view 😁 I've asked @garethbowen last week and we're going to keep access modifiers as it was originally, but I forgot the controllers, so this one is fixed, now revisiting the other functions. Thanks!

[the remark this answers is not in the document]
Comment (latin-panda) [reply webapp/src/ts/modules/messages/messages.component.ts:99]: Fixed.

File: webapp/src/ts/modules/messages/messages.component.ts (hunk 9 above)
Comment (latin-panda) [reply webapp/src/ts/modules/messages/messages.component.ts:62]: Fixed.

File: webapp/src/ts/modules/messages/messages.component.ts (hunk 14 above)
[the remark this answers is not in the document]
Comment (dianabarsan) [reply webapp/src/ts/modules/messages/messages.component.ts:99]: I think, ideally, tests should cover functionality fully, not implementation. A small refactor shouldn't warrant a unit test change, so testing "private" functions shouldn't really be necessary.

File: webapp/src/ts/modules/messages/messages.component.ts [hunk 15]
```diff
@@ -0,0 +1,125 @@
...
+          match.read = false;
+        }
+      } else {
+        currentConversations.push(updated);
+      }
+    });
+  }
+
+  listTrackBy(index, message) {
+    return message.key + message.doc._rev;
```
[the remark this answers is not in the document]
Comment (latin-panda) [reply webapp/src/ts/modules/messages/messages.component.ts:123]: 😱 so we can have docs with same rev? This is fixed.

File: webapp/src/ts/services/hydrate-messages.service.ts [hunk 16]
```diff
@@ -0,0 +1,99 @@
...
+      if (row.value.contact) {
+        contactIds.push(row.value.contact);
+      }
+    });
+
+    return this.lineageModelGenerator
+      .reportSubjects(contactIds)
+      .then((reports) => {
+        reports.forEach((report) => {
+          if (rowsObject[report._id]) {
```
[the remark this answers is not in the document]
Comment (latin-panda) [reply webapp/src/ts/services/hydrate-messages.service.ts:89]: I'm not very familiar with the hydrate code, but looks like it's not doing anything... can we remove `this.lineageModelGenerator.reportSubjects(contactIds)` and we just build the model of the rows `this.buildMessageModel(row.doc...`?

File: webapp/src/ts/services/message-contact.service.ts [hunk 17]
```diff
@@ -0,0 +1,110 @@
...
+    });
+  }
+
+  /**
+   * Update message view models
+   * @memberof AddReadStatus
+   * @param {Object[]} models The models to mark as read
+   * @returns {Promise} A Promise to return updated models
+   */
+  private addReadStatus(models = []) {
```
[the remark this answers is not in the document]
Comment (latin-panda) [reply webapp/src/ts/services/message-contact.service.ts:74]: Oh I thought reports didn't need it anymore because liveList got removed

File: webapp/src/ts/services/hydrate-messages.service.ts (hunk 16 above)
[the remark this answers is not in the document]
Comment (dianabarsan) [reply webapp/src/ts/services/hydrate-messages.service.ts:89]: I was only talking about the `rowsObject`, lineageModelGenerator is still necessary I think.

File: webapp/src/ts/modules/messages/messages.component.html [hunk 18]
```diff
@@ -0,0 +1,40 @@
...
+              <div class="summary"><p>{{ msg.message }}</p></div>
+              <div class="detail" *ngIf="msg.lineage && msg.lineage.length" [innerHTML]="msg.lineage | lineage"></div>
+            </div>
+          </a>
+        </li>
+      </ul>
+
+      <p *ngIf="error" class="alert alert-danger" role="alert" translate>Error fetching messages</p>
+      <p class="loading-status" *ngIf="!error && !loading && !conversations.length" translate>No messages found</p>
+      <p class="loading-status" *ngIf="!error && !loading && conversations.length">{{ 'message.list.complete' | translate }}</p>
```
[the remark this answers is not in the document]
Comment (latin-panda) [reply webapp/src/ts/modules/messages/messages.component.html:33]: I changed to use the directive. Besides format, is there any other advantage of the directive one over the pipe?

File: webapp/src/ts/services/message-contact.service.ts (hunk 17 above)
[the remark this answers is not in the document]
Comment (dianabarsan) [reply webapp/src/ts/services/message-contact.service.ts:74]: um ... live list has nothing to do with read status?

File: webapp/src/ts/modules/messages/messages.component.html (hunk 18 above)
[the remark this answers is not in the document]
Comment (dianabarsan) [reply webapp/src/ts/modules/messages/messages.component.html]: I don't think there are advantages, but it's just how all the other templates are doing it and I guess it's the "standard" :)

File: webapp/src/ts/modules/messages/messages.component.ts (hunk 15 above)
[the remark this answers is not in the document]
Comment (dianabarsan) [reply webapp/src/ts/modules/messages/messages.component.ts]: Revs are different than ids.

File: webapp/src/ts/services/hydrate-messages.service.ts (hunk 16 above)
[the remark this answers is not in the document]
Comment (dianabarsan) [reply webapp/src/ts/services/hydrate-messages.service.ts:89]: Ok ... this code is pretty hard to follow just by reading.  
`rowsObject` needs to stay. We're assigning `row` entities to `rowsObject`, and then mutating them and then using them when passing to `buildMessageModel` (via `row.report`).

File: webapp/src/ts/services/message-contact.service.ts (hunk 17 above)
[the remark this answers is not in the document]
Comment (latin-panda) [reply webapp/src/ts/services/message-contact.service.ts:74]: This is fixed.

File: webapp/src/ts/modules/messages/messages.component.html [hunk 19]
```diff
@@ -0,0 +1,40 @@
...
+  </div>
+</div>
+
+<div class="inner">
+  <div class="inbox page">
+
+    <div id="message-list" class="col-sm-4 inbox-items left-pane">
+      <ul *ngIf="!loading">
+        <li class="content-row"
+          *ngFor="let msg of conversations; trackBy: listTrackBy"
```
[the remark this answers is not in the document]
Comment (latin-panda) [reply webapp/src/ts/modules/messages/messages.component.html:13]: I might need this or something similar to load by default the first conversation in the list. I was planning to add this in the next PR about loading content. :)

File: webapp/src/ts/modules/messages/messages.component.ts [hunk 20]
```diff
@@ -0,0 +1,126 @@
...
+      this.store.pipe(select(Selectors.getLoadingContent)),
+      this.store.pipe(select(Selectors.getMessagesError)),
+    ).subscribe(([
+      conversations,
+      selectedConversation,
+      loadingContent,
+      error,
+    ]) => {
+      this.selectedConversation = selectedConversation;
+      this.conversations = (conversations || []).sort((a, b) => b.date - a.date);
```
Comment (dianabarsan) [inline webapp/src/ts/modules/messages/messages.component.ts:43]: I would prefer it if we do this in the reducer. It will also probably solve the error that this causes.

Comment (latin-panda) [reply webapp/src/ts/modules/messages/messages.component.ts]: Fixed

File: webapp/src/ts/modules/messages/messages.component.html (hunk 19 above)
[the remark this answers is not in the document]
Comment (dianabarsan) [reply webapp/src/ts/modules/messages/messages.component.html:13]: I don't think you need it for loading the 1st message anymore. 
It _was_ used before because we relied on the view sort filter to sort messages 🤯, but now we sort messages before passing to the template . 
It might be used by something else? Unsure.

**Review by dianabarsan — CHANGES_REQUESTED** [review]
Ok, this was the last pass and I found a few really minor things. Nice work!

File: webapp/src/ts/selectors/index.ts [hunk 21]
```diff
@@ -22,7 +23,13 @@ export const Selectors = {
   getReportsList: (state) => getReportsState(state).reports,
   listContains: (state) => (id) => getReportsState(state).reportsById.has(id),
   getSelectedReports: (state) => getReportsState(state).selected,
-}
+
+  // messages
+  getMessagesState: state => getMessagesState(state).messages,
```
[the remark this answers is not in the document]
Comment (latin-panda) [reply webapp/src/ts/selectors/index.ts:28]: ah yeah, it's removed.

File: webapp/tests/karma/ts/reducers/messages.spec.ts [hunk 22]
```diff
@@ -0,0 +1,61 @@
...
+    const expectedState = {
+      conversations: [{id: '124'}, {id: '567'}],
+      error: false,
+      messages: [],
+      selected: null
+    };
+
+    const result = messagesReducer(state, action);
+
+    expect(result).to.deep.include(expectedState);
```
[the remark this answers is not in the document]
Comment (latin-panda) [reply webapp/tests/karma/ts/reducers/messages.spec.ts:59]: It's added

File: webapp/tests/karma/ts/services/message-contact.service.spec.ts [hunk 23]
```diff
@@ -0,0 +1,292 @@
...
+          expect(getDataRecordsService.get.args[0]).to.deep.equal([['id1', 'id2', 'id3', 'id4'], { include_docs: true }]);
+          expect(hydrateMessagesService.hydrate.callCount).to.equal(1);
+          expect(hydrateMessagesService.hydrate.args[0]).to.deep.equal([[
+            { id: 'some_id1', value: { id: 'id1' }, doc: { _id: 'id1' } },
+            { id: 'some_id2', value: { id: 'id2' }, doc: { _id: 'id2' } },
+            { id: 'some_id3', value: { id: 'id3' }, doc: { _id: 'id3' } },
+            { id: 'some_id4', value: { id: 'id4' }, doc: { _id: 'id4' } },
+          ]]);
+          expect(list).to.deep.equal([
+            { id: 'some_id1', hydrated: true, read: true },
```
[the remark this answers is not in the document]
Comment (latin-panda) [reply webapp/tests/karma/ts/services/message-contact.service.spec.ts:85]: So, it's read status incorrect?

File: webapp/tests/karma/ts/services/message-contact.service.spec.ts [hunk 24]
```diff
@@ -0,0 +1,292 @@
...
+          expect(result).to.deep.equal([]);
+        });
+    });
+
+    it('should return errors from db query', () => {
+      dbService.get = () => ({
+        query: sinon.stub().rejects({ message: 'server error' })
+      });
+
+      service
```
[the remark this answers is not in the document]
Comment (latin-panda) [reply webapp/tests/karma/ts/services/message-contact.service.spec.ts:251]: Added.

File: webapp/tests/karma/karma-unit.conf.js [hunk 25]
```diff
@@ -1,52 +1,41 @@
...
-      'node_modules/moment/moment.js',
-      'node_modules/redux/dist/redux.js',
-      'node_modules/ng-redux/umd/ng-redux.js',
-      'node_modules/redux-thunk/dist/redux-thunk.js',
-      'tests/karma/utils.js',
-
-      // test files
-      'tests/karma/unit/**/*.js'
-    ]
+      { pattern: 'node_modules/sinon/pkg/sinon.js', instrument: false },
```
[the remark this answers is not in the document]
Comment (latin-panda) [reply webapp/tests/karma/karma-unit.conf.js]: Good catch! Fixed. I think this mapping isn't needed anymore and I removed it. Looks like unit test still running fine. Pls let me know if issue still persist.

File: webapp/tests/karma/ts/services/message-contact.service.spec.ts (hunk 23 above)
[the remark this answers is not in the document]
Comment (latin-panda) [reply webapp/tests/karma/ts/services/message-contact.service.spec.ts:85]: Looking at the code that sets this value in AddReadStatusService, appears to be expected 🤔
models.forEach((model, i) => {
    const row = response.rows[i];
    model.read = !!(row.value && !row.value.deleted); // doc exists.
});

[the remark this answers is not in the document]
Comment (latin-panda) [reply webapp/tests/karma/ts/services/message-contact.service.spec.ts:85]: It's a bit concerning having one array to access another by the same positioning idx, looks like it's working, but could easily break. Maybe something to improve in another task 🤔

[the remark this answers is not in the document]
Comment (dianabarsan) [reply webapp/tests/karma/ts/services/message-contact.service.spec.ts:85]: Yea, but you're using the same mock for the allDocs request that you're using for getting messages to getting read docs. 
We should either mock the meta db alldocs separately or mock the read docs service.

[the remark this answers is not in the document]
Comment (latin-panda) [reply webapp/tests/karma/ts/services/message-contact.service.spec.ts]: Ah because both services where using the same mocked data of DB Service 👍. So, now I'm mocking AddReadStatus Service as it was originally, it returns what it receives.

File: webapp/tests/karma/ts/reducers/messages.spec.ts (hunk 22 above)
[the remark this answers is not in the document]
Comment (dianabarsan) [reply webapp/tests/karma/ts/reducers/messages.spec.ts:59]: Where?

[the remark this answers is not in the document]
Comment (latin-panda) [reply webapp/tests/karma/ts/reducers/messages.spec.ts:59]: Line 48, added the property where it sorts by date timestamp and in line 51 the expected order. But shouldn't the controller decide how to display the data, including the if it's sorted or not? 🤔

[the remark this answers is not in the document]
Comment (dianabarsan) [reply webapp/tests/karma/ts/reducers/messages.spec.ts:59]:
````
const data = [{id: '124'}, {id: '567'}];
    const action = Actions.setConversations(data);
    const expectedState = {
      conversations: [{id: '124'}, {id: '567'}],
      error: false,
      messages: [],
      selected: null
    };
```

These are the lines you're referring to. I don't see any timestamp here.

[the remark this answers is not in the document]
Comment (dianabarsan) [reply webapp/tests/karma/ts/reducers/messages.spec.ts:59]: Sorry, for some reason, github did not display the change about adding the date. it still showed the old code when I checked the file. :( My apologies!

**Review by dianabarsan — APPROVED** [review]
🚀
